File: src/test/kotlin/com/tambapps/http/restclient/AbstractRestClientTest.java
package com.tambapps.http.restclient;

import static org.junit.Assert.*;

import com.google.gson.reflect.TypeToken;
import com.tambapps.http.restclient.response.handler.ResponseHandler;
import com.tambapps.http.restclient.response.handler.ResponseHandlers;
import com.tambapps.http.restclient.response.RestResponse;
import com.tambapps.http.restclient.util.ObjectListParser;
import com.tambapps.http.restclient.util.ObjectParser;

import com.google.gson.Gson;

import java.lang.reflect.Type;
import java.util.List;
import java.util.Objects;

public abstract class AbstractRestClientTest {

  static final String API_URL = "https://jsonplaceholder.typicode.com/";
  static final int TIMEOUT = 8;
  static final Gson GSON = new Gson();
  private static final ObjectParser JSON_PARSER = new ObjectParser() {
    @Override
    public <T> T parse(Class<T> clazz, String data) {
      return GSON.fromJson(data, clazz);
    }
  };
  static final ResponseHandler<Post> RESPONSE_HANDLER =
    ResponseHandlers.object(Post.class, JSON_PARSER);

  private static final Type POST_LIST_TYPE = new TypeToken<List<Post>>(){}.getType();
  private static final ObjectListParser JSON_LIST_PARSER = new ObjectListParser() {
    @Override
    public <T> List<T> parse(Class<T> clazz, String data) {
      return GSON.fromJson(data, POST_LIST_TYPE);
    }
  };
  static final ResponseHandler<List<Post>> LIST_RESPONSE_HANDLER =
    ResponseHandlers.objectList(Post.class, JSON_LIST_PARSER);

  void getAsserts(RestResponse<Post> response) {
    assertTrue("Should be successful", response.isSuccessful());
    assertFalse("Shouldn't be an error code", response.isErrorResponse());

    assertPost(response.data);
  }

  private void assertPost(Post post) {
    assertNotNull("Shouldn't be null", post);
    assertNotNull("Shouldn't be null", post.getId());
    assertNotNull("Shouldn't be null", post.getUserId());
    assertNotNull("Shouldn't be null", post.getBody());
    assertNotNull("Shouldn't be null", post.getTitle());
  }
  void getListAsserts(RestResponse<List<Post>> response) {
    assertTrue("Should be successful", response.isSuccessful());
    assertFalse("Shouldn't be an error code", response.isErrorResponse());

    for (Post post : response.data) {
      assertPost(post);
    }
  }

  void getListAssertsWithUserId(RestResponse<List<Post>> response, int userId) {
    getListAsserts(response);
    for (Post post : response.data) {
      assertEquals("Should be equal", userId, (int) post.getUserId());
    }
  }

  void putAsserts(RestResponse<Post> response, Post post) {
    assertFalse("Shouldn't be an error code", response.isErrorResponse());
    assertTrue("Should be successful", response.isSuccessful());
    assertEquals("Should be equal", post, response.data);
  }

  void postAsserts(RestResponse<Post> response, Post post) {
    assertTrue("Should be successful", response.isSuccessful());
    assertFalse("Shouldn't be null", response.isErrorResponse());
    Post responseData = response.data;
    assertEquals("Should be equal", post.getUserId(), responseData.getUserId());
    assertEquals("Should be equal", post.getTitle(), responseData.getTitle());
    assertEquals("Should be equal", post.getBody(), responseData.getBody());
  }

  void deleteAsserts(RestResponse<Post> response) {
    assertTrue("Should be successful", response.isSuccessful());
    assertFalse("Shouldn't be an error code", response.isErrorResponse());
  }

  protected static class Post {
    private Integer id;
    private Integer userId;
    private String title;
    private String body;

    public Post(Integer id, Integer userId, String title, String body) {
      this.id = id;
      this.userId = userId;
      this.title = title;
      this.body = body;
    }

    public Integer getId() {
      return id;
    }

    public Integer getUserId() {
      return userId;
    }

    public String getTitle() {
      return title;
    }

    public String getBody() {
      return body;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      Post post = (Post) o;
      return Objects.equals(id, post.id) &&
        Objects.equals(userId, post.userId) &&
        Objects.equals(title, post.title) &&
        Objects.equals(body, post.body);
    }

    @Override
    public int hashCode() {
      return Objects.hash(id);
    }

    @Override
    public String toString() {
      return "Post{" +
        "id=" + id +
        ", userId=" + userId +
        ", title='" + title + '\'' +
        ", body='" + body + '\'' +
        '}';
    }
  }
}

File: src/test/kotlin/com/tambapps/http/client/AbstractRestClientTest.kt
package com.tambapps.http.client

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.tambapps.http.client.response.RestResponse
import com.tambapps.http.client.response.handler.ResponseHandlers.`object`
import com.tambapps.http.client.response.handler.ResponseHandlers.objectList
import com.tambapps.http.client.util.ObjectListParser
import com.tambapps.http.client.util.ObjectParser
import junit.framework.Assert.assertEquals
import org.junit.Assert
import java.util.*

abstract class AbstractRestClientTest {
    fun getAsserts(response: RestResponse<Post>) {
        Assert.assertTrue("Should be successful", response.isSuccessful)
        Assert.assertFalse("Shouldn't be an error code", response.isErrorResponse)
        assertPost(response.data)
    }

    private fun assertPost(post: Post) {
        Assert.assertNotNull("Shouldn't be null", post)
        Assert.assertNotNull("Shouldn't be null", post.id)
        Assert.assertNotNull("Shouldn't be null", post.userId)
        Assert.assertNotNull("Shouldn't be null", post.body)
        Assert.assertNotNull("Shouldn't be null", post.title)
    }

    fun getListAsserts(response: RestResponse<List<Post>>) {
        Assert.assertTrue("Should be successful", response.isSuccessful)
        Assert.assertFalse("Shouldn't be an error code", response.isErrorResponse)
        for (post in response.data) {
            assertPost(post)
        }
    }

    fun getListAssertsWithUserId(response: RestResponse<List<Post>>, userId: Int) {
        getListAsserts(response)
        for (post in response.data) {
            Assert.assertEquals("Should be equal", userId, post.userId)
        }
    }

    fun putAsserts(response: RestResponse<Post?>, post: Post?) {
        Assert.assertFalse("Shouldn't be an error code", response.isErrorResponse)
        Assert.assertTrue("Should be successful", response.isSuccessful)
        assertEquals("Should be equal", post, response.data)
    }

    fun postAsserts(response: RestResponse<Post>, post: Post) {
        Assert.assertTrue("Should be successful", response.isSuccessful)
        Assert.assertFalse("Shouldn't be null", response.isErrorResponse)
        val responseData = response.data
        Assert.assertEquals("Should be equal", post.userId, responseData.userId)
        Assert.assertEquals("Should be equal", post.title, responseData.title)
        Assert.assertEquals("Should be equal", post.body, responseData.body)
    }

    fun deleteAsserts(response: RestResponse<Post?>) {
        Assert.assertTrue("Should be successful", response.isSuccessful)
        Assert.assertFalse("Shouldn't be an error code", response.isErrorResponse)
    }

    class Post(val id: Int, val userId: Int, val title: String, val body: String) {

        override fun equals(o: Any?): Boolean {
            if (this === o) {
                return true
            }
            if (o == null || javaClass != o.javaClass) {
                return false
            }
            val post = o as Post
            return id == post.id &&
                    userId == post.userId &&
                    title == post.title &&
                    body == post.body
        }

        override fun hashCode(): Int {
            return Objects.hash(id)
        }

        override fun toString(): String {
            return "Post{" +
                    "id=" + id +
                    ", userId=" + userId +
                    ", title='" + title + '\'' +
                    ", body='" + body + '\'' +
                    '}'
        }

    }

    companion object {
        const val API_URL = "https://jsonplaceholder.typicode.com/"
        const val TIMEOUT = 8
        @JvmField
        val GSON = Gson()
        private val JSON_PARSER: ObjectParser = object : ObjectParser {
            override fun <T> parse(clazz: Class<T>, data: String): T {
                return GSON.fromJson(data, clazz)
            }
        }
        @JvmField
        val RESPONSE_HANDLER = `object`(Post::class.java, JSON_PARSER)
        private val POST_LIST_TYPE = object : TypeToken<List<Post?>?>() {}.type
        private val JSON_LIST_PARSER: ObjectListParser = object : ObjectListParser {
            override fun <T> parse(clazz: Class<T>, data: String): List<T> {
                return GSON.fromJson(data, POST_LIST_TYPE)
            }
        }
        @JvmField
        val LIST_RESPONSE_HANDLER = objectList(Post::class.java, JSON_LIST_PARSER)
    }
}